crest/Assets/Crest/Crest/Scripts/Helpers/FloatingOrigin.cs

```diff
@@ -65,35 +69,149 @@ public class FloatingOrigin : MonoBehaviour
+            while (true)
+            {
+                // NOTE: Will not work in batch mode:
+                // https://docs.unity3d.com/Manual/CLIBatchmodeCoroutines.html
+                yield return Helpers.WaitForEndOfFrame;
+
+                if (_debug._isCapturing)
+                {
+                    ExternalGPUProfiler.EndGPUCapture();
+                    _debug._isCapturing = false;
+                }
+
+#if UNITY_EDITOR
+                if (_debug._pauseOnShift && TeleportOriginThisFrame != Vector3.zero)
+                {
+                    UnityEditor.EditorApplication.isPaused = true;
+                }
+#endif
+
+                // Cannot clear at start of FixedUpdate (without extra logic) as it can run multiple times.
+                TeleportOriginThisFrame = Vector3.zero;
 
+                _isFirstFixedUpdate = true;
+            }
+        }
+
+        void FixedUpdate()
+        {
+            if (!_isFirstFixedUpdate)
+            {
+                return;
+            }
+
+            _isFirstFixedUpdate = false;
             var newOrigin = Vector3.zero;
-            if (Mathf.Abs(transform.position.x) > _threshold) newOrigin.x += transform.position.x;
-            if (Mathf.Abs(transform.position.z) > _threshold) newOrigin.z += transform.position.z;
+
+            if (Mathf.Abs(transform.position.x) > _threshold)
+            {
+#if UNITY_EDITOR
+                if (_debug._pauseBeforeShift && !_debug._shiftNextX)
+                {
+                    _debug._shiftNextX = true;
+                    UnityEditor.EditorApplication.isPaused = true;
+                }
+                else
+#endif // UNITY_EDITOR
+                {
+                    // Teleport by threshold value intervals to avoid popping.
+                    newOrigin.x += Mathf.Floor(transform.position.x / _threshold) * _threshold;
+#if UNITY_EDITOR
+                    _debug._shiftNextX = false;
+#endif
+                }
+            }
+
+            if (Mathf.Abs(transform.position.z) > _threshold)
+            {
+#if UNITY_EDITOR
+                if (_debug._pauseBeforeShift && !_debug._shiftNextY)
+                {
+                    _debug._shiftNextY = true;
+                    UnityEditor.EditorApplication.isPaused = true;
+                }
+                else
+#endif // UNITY_EDITOR
+                {
+                    // Teleport by threshold value intervals to avoid popping.
+                    newOrigin.z += Mathf.Floor(transform.position.z / _threshold) * _threshold;
+#if UNITY_EDITOR
+                    _debug._shiftNextY = false;
+#endif
+                }
+            }
 
             if (newOrigin != Vector3.zero)
             {
                 MoveOrigin(newOrigin);
             }
         }
 
+        void OnValidate()
+        {
+            // Must be power of two to avoid popping.
+            _threshold = Mathf.Pow(2f, Mathf.Round(Mathf.Log(_threshold, 2f)));
```

Comment:
does > 2048 need to be validated here too? I know you have a min attribute but its public and could be set from code or users could hack the min i guess

Reply:
Good find. Done. Thanks.